numpy/core/src/private/mem_overlap.c

```diff
@@ -0,0 +1,814 @@
+/*
+  Solving memory overlap integer programs and bounded Diophantine equations with
+  positive coefficients.
+
+  Asking whether two strided arrays `a` and `b` overlap is equivalent to
+  asking whether there is a solution to the following problem::
+
+      sum(stride_a[i] * x_a[i] for i in range(ndim_a))
+      -
+      sum(stride_b[i] * x_b[i] for i in range(ndim_b))
+      ==
+      base_b - base_a
+
+      0 <= x_a[i] <= shape_a[i]
+      0 <= x_b[i] <= shape_b[i]
+
+  for some integer x_a, x_b.  Itemsize needs to be considered as an additional
+  dimension with stride 1 and size itemsize-1.
```

Comment:
The comment is wrong here: shape_a[i] is exclusive upper bound, and the
size is itemsize-1. The code internally uses inclusive upper bounds,
because that's how it's done in all papers.